toolchain/lower/file_context.cpp

```diff
@@ -377,12 +351,17 @@ auto FileContext::BuildFunctionDefinition(SemIR::FunctionId function_id)
   // The subset of calling_convention_param_id that is in sequential order.
   llvm::ArrayRef<SemIR::InstId> sequential_param_ids =
       calling_convention_param_ids;
+  if (function.return_slot_id.is_valid()) {
+    sequential_param_ids = sequential_param_ids.drop_back();
+  }
 
   // The LLVM calling convention has the return slot first rather than last.
-  if (has_return_slot) {
+  // Note that this queries whether there is a return slot at the LLVM level,
+  // whereas `function.return_slot_id.is_valid()` queries whether there is a
+  // return slot at the return slot at the SemIR level.
+  if (SemIR::ReturnTypeInfo::ForFunction(sem_ir(), function, specific_id)
+          .has_return_slot()) {
```

Comment:
It's interesting to me that these two `if` conditions are separate. Are there cases where:

1. `return_slot_id.is_valid()` returns false, but `has_return_slot` returns true?
2. `has_return_slot` returns false, but `return_slot_id.is_valid()` returns true?

If one case exists and the other does not, perhaps the `if` should be nested?

Given the addition of `calling_convention_param_ids`, maybe you could replace `sequential_param_ids` and write this as something like:

```
if (return_slot_id.is_valid()) {
  // If necessary, you could write `if (has_return_slot)` here
  lower_param(calling_convention_param_ids.back());
  calling_convention_param_ids = calling_convention_param_ids.drop_back();
}
```

Reply:
> It's interesting to me that these two `if` conditions are separate. Are there cases where:
> 
> 1. `return_slot_id.is_valid()` returns false, but `has_return_slot` returns true?
> 2. `has_return_slot` returns false, but `return_slot_id.is_valid()` returns true?
> 
> If one case exists and the other does not, perhaps the `if` should be nested?

I believe 2 is possible but 1 is not, so I've nested the conditions.

> 
> Given the addition of `calling_convention_param_ids`, maybe you could replace `sequential_param_ids` and write this as something like:
> 
> ```
> if (return_slot_id.is_valid()) {
>   // If necessary, you could write `if (has_return_slot)` here
>   lower_param(calling_convention_param_ids.back());
>   calling_convention_param_ids = calling_convention_param_ids.drop_back();
> }
> ```

I considered that, but decided not to because it could be confusing to have a variable named `calling_convention_param_ids` that doesn't actually contain the full list of calling convention params. I could do it the other way around by dropping `calling_convention_param_ids` and using `sequential_param_ids` everywhere, but that has the same problem: `sequential_param_ids` wouldn't be in sequential order until the end of this branch (and the code within the branch would be relying on that fact).